storage: fix read behavior for large key-values

sumeerbhola opened the pull request:
pebbleMVCCScanner no longer has a 128MB limit on the size
of the key-value pair it can read.

Release justification: Fix for high-severity bug in existing
functionality. The pebbleMVCCScanner change prevents a panic
if a key-value pair larger than 128MB gets written, which
affected a customer cluster.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61872)
<!-- Reviewable:end -->

sumeerbhola reviewed (commented):
I'll add some testing, but I am somewhat concerned about the fallout from enforcing on the write path, as mentioned in the code comment, and would like your opinion.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-:-MVYGJme7RQoU2B7H8T0:b-vuosye)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @itsbilal, and @petermattis)



<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @itsbilal, and @sumeerbhola)

___
*[pkg/storage/engine.go, line 1034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaaZ4qCyZcXexQHGg7:-MVaaZ4qCyZcXexQHGg8:bu5p9vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/engine.go#L1034)):*
> ```Go
> // imposes a stricter limit than this code. However, a cluster setting can be
> // increased and we do not want to allow huge key-value pairs to creep into
> // CockroachDB, due to the negative performance implications. This code must
> ```

I thought there was a mechanism to limit the values of cluster settings.
___
*[pkg/storage/engine.go, line 1047 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaah1w8VmgKVwkDSGB:-MVaah1w8VmgKVwkDSGC:b-oxyteg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/engine.go#L1047)):*
> ```Go
> // table/index were dropped).
> func checkMVCCWriteSizeLimit(key MVCCKey, value []byte) error {
> 	const maxMVCCWriteSize = 64 << 20
> ```

I'm quite worried about this being hardcoded given the comment above. 
___
*[pkg/storage/engine.go, line 1049 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaaEgNA5PjyZgCw_ww:-MVaaEgO2yLxllhPOxbg:b-z21c7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/engine.go#L1049)):*
> ```Go
> 	const maxMVCCWriteSize = 64 << 20
> 	if key.Len()+len(value) > maxMVCCWriteSize {
> 		return errors.Errorf("key-value pair is too large")
> ```

The error message should probably contain the size of the key and value and the limit that was exceeded.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaamdVB4Zkc5LBBLRO:-MVaamdVB4Zkc5LBBLRP:b-6z4f4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/pebble_mvcc_scanner.go#L69)):*
> ```Go
> 			newSize *= 2
> 		}
> 		if newSize < lenToAdd {
> ```

This is safe and good. Perhaps it should be pulled into a separate change. I can see including this in 21.1. I'd be quite anxious about including the KV size limit in 21.1 unless we have a good escape hatch. 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine.go, line 1034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaaZ4qCyZcXexQHGg7:-MVdH1_s8fDSB5X28IDr:b-vp859e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/engine.go#L1034)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I thought there was a mechanism to limit the values of cluster settings.
</blockquote></details>

There is a `MaxCommandSizeFloor` minimum for this setting, but no max. Adding a max would probably suffer from the same concerns, as the code comment below, of what happens to existing clusters that have used a value larger than the max.
___
*[pkg/storage/engine.go, line 1047 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaah1w8VmgKVwkDSGB:-MVdHTeN8O_R0YhqOSzM:bjo49pu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/engine.go#L1047)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm quite worried about this being hardcoded given the comment above. 
</blockquote></details>

I'll get rid of this then.

@andreimatei do you have any thoughts?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> Writer.PutMVCC imposes a size limit on the
> versioned key-value pairs being written, so that we are
> not wholly reliant on kv.raft.command.max_size for
> enforcement.

If the reader can now deal with large values, then why do we need more protection than ` kv.raft.command.max_size`? Particularly since Pebble seems able to deal with larger values since you say that the new limit doesn't apply to meta keys.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-:-MVqYI2K2phU20Z55_-z:bcfoflk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine.go, line 1034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaaZ4qCyZcXexQHGg7:-MVqXbEH-d_mIPbJX4CE:bq1kltu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/engine.go#L1034)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

There is a `MaxCommandSizeFloor` minimum for this setting, but no max. Adding a max would probably suffer from the same concerns, as the code comment below, of what happens to existing clusters that have used a value larger than the max.
</blockquote></details>

I believe we can attach a validation function to a cluster setting, and it seems that we should do that if this code hardcodes 64MB.


<!-- Sent from Reviewable.io -->

sumeerbhola committed:
storage: fix read behavior for large key-values

pebbleMVCCScanner no longer has a 128MB limit on the size
of the key-value pair it can read.

Release justification: Fix for high-severity bug in existing
functionality. The pebbleMVCCScanner change prevents a panic
if a key-value pair larger than 128MB gets written, which
affected a customer cluster.

Release note: None

sumeerbhola reviewed (commented):
TFTRs!

I've added a test.

> If the reader can now deal with large values, then why do we need more protection than kv.raft.command.max_size? Particularly since Pebble seems able to deal with larger values since you say that the new limit doesn't apply to meta keys.

The idea was that we don't want large values, since they are poorly handled from an efficiency perspective. The fact that we can't enforce more strictly because of history in intent meta should not detract from doing what we can. 

Since this is controversial, given we have no control on what has already happened in customer clusters, I have removed it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-:-MVvI7Gb3-Z9uG_AZlLf:b-ib8iut)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @itsbilal, and @petermattis)

___
*[pkg/storage/engine.go, line 1034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaaZ4qCyZcXexQHGg7:-MVvTsWF6Nzs90jZmTgC:bnaga1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/engine.go#L1034)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I believe we can attach a validation function to a cluster setting, and it seems that we should do that if this code hardcodes 64MB.
</blockquote></details>

I've removed the write path enforcement, so no need for fiddling with validation functions.
___
*[pkg/storage/engine.go, line 1049 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaaEgNA5PjyZgCw_ww:-MVvU1hoAYnvfhv0_Ftm:ba3dfob) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/engine.go#L1049)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The error message should probably contain the size of the key and value and the limit that was exceeded.
</blockquote></details>

Removed
___
*[pkg/storage/pebble_mvcc_scanner.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVaamdVB4Zkc5LBBLRO:-MVvToKbCJAuK3VW_tsy:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e67ec6702b6122bbae6ce8df4ea0c3d551698a5/pkg/storage/pebble_mvcc_scanner.go#L69)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is safe and good. Perhaps it should be pulled into a separate change. I can see including this in 21.1. I'd be quite anxious about including the KV size limit in 21.1 unless we have a good escape hatch. 
</blockquote></details>

Done
___
*[pkg/storage/pebble_mvcc_scanner.go, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-MVvTG2rEcrAVQ9Os-SC:-MVvTG2rEcrAVQ9Os-SD:b-hf1axa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2de862286a8e2b5fd0d38767d5e3ece66620fca6/pkg/storage/pebble_mvcc_scanner.go#L67)):*
> ```Go
> 			// for next allocation, and instead reset the exponential increase, in
> 			// case we had a stray huge key-value.
> 			newSize = minSize
> ```

I've changed the logic here to not double from a previous huge key-value. The original change caused the test to allocate `314572838` when it reached `keys[2]`. Because of this reset, the code below also avoids the for-loop if the `lenToAdd` is known to be huge.


<!-- Sent from Reviewable.io -->

sumeerbhola commented:
any more comments? I'd like to merge this soon.

andreimatei reviewed (commented):
not from me, but I'm not the right reviewer for the changed code

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-:-MW9s0PuCLSIpplXmdsf:b-4u7urq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @itsbilal, and @petermattis)



<!-- Sent from Reviewable.io -->

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 5 files at r1, 3 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-:-MWA3osT6sYKxTMFiHVH:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @petermattis)



<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61872#-:-MWAaA4qAE_VEcAx7AYt:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @petermattis)



<!-- Sent from Reviewable.io -->

sumeerbhola commented:
bors r+
